### ISSUE TITLE
Sphinx 4.1 fails to parse advanced C++ template usage

### ISSUE THREAD

**Issue by Luthaf** [sphinx-doc/sphinx#9477]
### Describe the bug

Sphinx 4.1 fails when parsing some advanced uses of C++ templates (using the breathe extension):

```cpp
template<typename... T> void parameter_pack(T... u) {}
```

Gives this error:

<details>

```
# Sphinx version: 4.1.0
# Python version: 3.9.0 (CPython)
# Docutils version: 0.17.1 release
# Jinja2 version: 2.11.3
# Last messages:
#   Sphinx v4.1.0 en cours d'exécution
#   Chargement de l'environnement pickled...
#   échoué
#   échec : version non à jour de l’environnement de construction
#   Construction en cours [mo] : cibles périmées pour les fichiers po 0
#   Construction [html] : cibles périmées pour les fichiers sources 1
#   Mise à jour de l'environnement :
#   [nouvelle configuration]
#   1 ajouté(s), 0 modifié(s), 0 supprimé(s)
#   Lecture des sources... [100%] index
# Loaded extensions:
#   sphinx.ext.mathjax (4.1.0) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (2.0.0) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.5) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/alabaster/__init__.py
#   breathe (4.30.0) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/__init__.py
Traceback (most recent call last):
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/application.py", line 343, in build
    self.builder.build_update()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 307, in build
    updated_docnames = set(self.read())
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 414, in read
    self._read_serial(docnames)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 435, in _read_serial
    self.read_doc(docname)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 475, in read_doc
    doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/io.py", line 188, in read_doc
    pub.publish()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/core.py", line 217, in publish
    self.document = self.reader.read(self.source, self.parser,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/io.py", line 108, in read
    self.parse()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/readers/__init__.py", line 78, in parse
    self.parser.parse(self.input, document)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/parsers.py", line 100, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 170, in run
    results = StateMachineWS.run(self, input_lines, input_offset,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 239, in run
    context, next_state, result = self.check_line(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 451, in check_line
    return method(match, context, next_state)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2769, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 393, in new_subsection
    newabsoffset = self.nested_parse(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 281, in nested_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 239, in run
    context, next_state, result = self.check_line(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 451, in check_line
    return method(match, context, next_state)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2342, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2354, in explicit_construct
    return method(self, expmatch)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2096, in directive
    return self.run_directive(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2146, in run_directive
    result = directive_instance.run()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/directives.py", line 110, in run
    node_stack = self._resolve_function(matches, args, project_info)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/directives.py", line 218, in _resolve_function
    signature = self._create_function_signature(entry, project_info, filter_,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/directives.py", line 188, in _create_function_signature
    object_renderer.create_template_prefix(node),
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 780, in create_template_prefix
    nodes = self.render(decl.templateparamlist)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2279, in render
    result = method(self, node)  # type: ignore
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2089, in visit_templateparamlist
    nodelist.extend(self.visit_param(item, insertDeclNameByParsing=True))
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2051, in visit_param
    ast = parser._parse_type(named='single', outer='templateParam')
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 6415, in _parse_type
    declSpecs = self._parse_decl_specs(outer=outer)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 6109, in _parse_decl_specs
    trailing = self._parse_trailing_type_spec()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 5874, in _parse_trailing_type_spec
    nestedName = self._parse_nested_name()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 5796, in _parse_nested_name
    self.fail("Expected identifier in nested name.")
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 275, in fail
    raise self._make_multi_error(errors, '')
sphinx.util.cfamily.DefinitionError: Invalid C++ declaration: Expected identifier in nested name. [error at 8]
  typename...
  --------^
```

</details>

And this code

```cpp
template <template <class> class HighOrder, typename T>
HighOrder<T> high_order_template() {}
```

Gives this error:

<details>

```
# Sphinx version: 4.1.0
# Python version: 3.9.0 (CPython)
# Docutils version: 0.17.1 release
# Jinja2 version: 2.11.3
# Last messages:
#   Sphinx v4.1.0 en cours d'exécution
#   Construction en cours [mo] : cibles périmées pour les fichiers po 0
#   Construction [html] : cibles périmées pour les fichiers sources 1
#   Mise à jour de l'environnement :
#   [nouvelle configuration]
#   1 ajouté(s), 0 modifié(s), 0 supprimé(s)
#   Lecture des sources... [100%] index
# Loaded extensions:
#   sphinx.ext.mathjax (4.1.0) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (2.0.0) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.5) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/alabaster/__init__.py
#   breathe (4.30.0) from ~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/__init__.py
Traceback (most recent call last):
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/application.py", line 343, in build
    self.builder.build_update()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 307, in build
    updated_docnames = set(self.read())
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 414, in read
    self._read_serial(docnames)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 435, in _read_serial
    self.read_doc(docname)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 475, in read_doc
    doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/io.py", line 188, in read_doc
    pub.publish()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/core.py", line 217, in publish
    self.document = self.reader.read(self.source, self.parser,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/io.py", line 108, in read
    self.parse()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/readers/__init__.py", line 78, in parse
    self.parser.parse(self.input, document)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/parsers.py", line 100, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 170, in run
    results = StateMachineWS.run(self, input_lines, input_offset,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 239, in run
    context, next_state, result = self.check_line(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 451, in check_line
    return method(match, context, next_state)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2769, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 393, in new_subsection
    newabsoffset = self.nested_parse(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 281, in nested_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 239, in run
    context, next_state, result = self.check_line(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 451, in check_line
    return method(match, context, next_state)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2344, in explicit_markup
    self.explicit_list(blank_finish)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2369, in explicit_list
    newline_offset, blank_finish = self.nested_list_parse(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 318, in nested_list_parse
    state_machine.run(block, input_offset, memo=self.memo,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 239, in run
    context, next_state, result = self.check_line(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/statemachine.py", line 451, in check_line
    return method(match, context, next_state)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2647, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2354, in explicit_construct
    return method(self, expmatch)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2096, in directive
    return self.run_directive(
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/docutils/parsers/rst/states.py", line 2146, in run_directive
    result = directive_instance.run()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/directives.py", line 110, in run
    node_stack = self._resolve_function(matches, args, project_info)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/directives.py", line 218, in _resolve_function
    signature = self._create_function_signature(entry, project_info, filter_,
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/directives.py", line 188, in _create_function_signature
    object_renderer.create_template_prefix(node),
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 780, in create_template_prefix
    nodes = self.render(decl.templateparamlist)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2279, in render
    result = method(self, node)  # type: ignore
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2089, in visit_templateparamlist
    nodelist.extend(self.visit_param(item, insertDeclNameByParsing=True))
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/breathe/renderer/sphinxrenderer.py", line 2051, in visit_param
    ast = parser._parse_type(named='single', outer='templateParam')
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 6415, in _parse_type
    declSpecs = self._parse_decl_specs(outer=outer)
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 6109, in _parse_decl_specs
    trailing = self._parse_trailing_type_spec()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 5874, in _parse_trailing_type_spec
    nestedName = self._parse_nested_name()
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/domains/cpp.py", line 5800, in _parse_nested_name
    self.fail("Expected identifier in nested name, "
  File "~/sphinx-c++/virtualenv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 275, in fail
    raise self._make_multi_error(errors, '')
sphinx.util.cfamily.DefinitionError: Invalid C++ declaration: Expected identifier in nested name, got keyword: template [error at 8]
  template< class > class
  --------^
```

</details>

----

Sphinx 4.0.1 is able to build the documentation without problems.

### How to Reproduce

I have a reproduction here: https://github.com/Luthaf/sphinx-4.1-parse-error

```
$ git clone https://github.com/Luthaf/sphinx-4.1-parse-error
$ cd sphinx-4.1-parse-error
$ pip install -r requirements.txt
$ make html
```

### Your project

https://github.com/Luthaf/sphinx-4.1-parse-error

### OS

macOS, we also see this on Linux in CI

### Python version

3.9

### Sphinx version

4.1.0

### Sphinx extensions

Breathe 4.30.0 with doxygen 1.9.1

**Comment by Luthaf** [comment]
#9451 might be related as well, but I could not reproduce in my tests.

**Comment by jakobandersen** [comment]
Thanks for reporting. I'm quite sure this is a duplicate of #9449, #9433, and #9460 though, but otherwise please reopen or post a new issue.
The problem is in Breathe and I think it should be fixed by michaeljones/breathe#711.

**Comment by Luthaf** [comment]
I don't know how I missed all these other issues, sorry for the noise. I can confirm that using the breathe branch fixes the issue. Thanks a lot for your work!
